src/gui/qgsmapcanvas.cpp

```diff
@@ -202,6 +202,7 @@ QgsMapCanvas::QgsMapCanvas( QWidget * parent, const char *name )
   mCurrentLayer = NULL;
   mMapOverview = NULL;
   mMapTool = NULL;
+  mOldCursor = cursor();
```

Comment:
Yes, you are right. I agree that keeping and restoring of the cursor has to
be done in the base class QgsMapTool.  I will modify it instead
qgsmapcanvas. Additionally I am planning to move  QgsMeasureTool in the
library. Now is in QGIS application.



On Tue, Jul 28, 2015 at 4:38 PM, Matthias Kuhn <notifications@github.com>
wrote:

> In src/gui/qgsmapcanvas.cpp
> <https://github.com/qgis/QGIS/pull/2224#discussion_r35647206>:
>
> > @@ -202,6 +202,7 @@ QgsMapCanvas::QgsMapCanvas( QWidget * parent, const char *name )
> >    mCurrentLayer = NULL;
> >    mMapOverview = NULL;
> >    mMapTool = NULL;
> > +  mOldCursor = cursor();
>
> I guess there is a conceptual problem about this:
>
> Due to the fact that maptools don’t restore cursor, the previous cursor
> shape is kept in mOldCursor.
>
> I don't see a reason why maptools should not restore the cursor (like
> everything else in the application does)
> Qt has a cursor stack:
>
>    - you want to change the cursor: push a change
>    - you want to restore the curstor: pop the change
>
> But since only the maptool knows if (and how many times) it pushed
> something onto the cursor stack only the maptool itself can pop that again.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/qgis/QGIS/pull/2224/files#r35647206>.
>